src/Analyser/ConstantResolver.php

```diff
@@ -41,6 +43,7 @@ public function __construct(
 		private array $dynamicConstantNames,
 		private ?PhpVersion $composerMinPhpVersion,
 		private ?PhpVersion $composerMaxPhpVersion,
+		private ?Container $container,
```

Comment:
Just make this new param the TypeStringRrsolver itself. 

It will be injected by the DIC automatically

Reply:
I initially thought that might be the way to do it. However, when I tried it, I wound up getting an error about circular dependencies. So that's why I switched to the entire container. Do you have any ideas on how to get around the circular dependency error? It's also possible I just did it wrong. I assumed I needed to update the places that called the constructor, so I was using the container there to get the TypeStringResolver.

Reply:
or maybe instead remove `TypeNodeResolver` from `TypeStringResolver` construction-time deps and move it to setter-injection

Reply:
Sorry, @staabm. I don't have experience with DI in PHP, so I'm not entirely sure how to do this. I assume you meant to add a `setTypeNodeResolver` function to `TypeStringResolver`. I'm not sure where I would call `setTypeNodeResolver` though. I thought maybe I needed to add a `TypeStringResolverFactory`. I'm guessing that maybe I need to update `conf/config.neon`, so I changed it to this:
```neon
-
	class: PHPStan\PhpDoc\TypeStringResolver
typeStringResolverFactory:
	class: PHPStan\PhpDoc\TypeStringResolverFactory
```

In a new `TypeStringResolverFactory` class, I added this:
```php
public function create(): TypeStringResolver
{
	$typeStringResolver = new TypeStringResolver(
		$this->typeLexer,
		$this->typeParser
	);
	$typeStringResolver->setTypeNodeResolver($this->container->getByType(TypeNodeResolver::class));

	return $typeStringResolver;
}
```

I still wound up with this when running the test though:
```
In Container.php line 327:

  Circular reference detected for: reflectionProvider, betterReflectionProvider, 093,
  058, 042, 040.
```

Am I on the right track? Do you have any suggestions? Thanks.